datastore/sqlite_datastore.nim

```diff
@@ -0,0 +1,272 @@
+  # https://www.sqlite.org/pragma.html#pragma_page_size
+  # https://www.sqlite.org/intern-v-extern-blob.html
+
+  var
+    env: AutoDisposed[SQLite]
+
+  defer: disposeIfUnreleased(env)
+
+  var
+    basep, fname, dbPath: string
+
+  if inMemory:
+    if readOnly:
+      return failure "SQLiteDatastore cannot be read-only and in-memory"
+    else:
+      dbPath = ":memory:"
+  else:
+    try:
+      basep = normalizePathEnd(
+        if basePath.isAbsolute: basePath
+        else: getCurrentDir() / basePath)
+
+      fname = filename.normalizePathEnd
+      dbPath = basep / fname
+
+      if readOnly and not fileExists(dbPath):
+        return failure "read-only database does not exist: " & dbPath
+      else:
+        createDir(basep)
+
+    except IOError as e:
+      return failure e
+
+    except OSError as e:
+      return failure e
+
+  let
+    flags =
+      if readOnly: SQLITE_OPEN_READONLY
+      else: SQLITE_OPEN_READWRITE or SQLITE_OPEN_CREATE
+
+  open(dbPath, env.val, flags)
+
+  let
+    pragmaStmt = journalModePragmaStmt(env.val)
+
+  checkExec(pragmaStmt)
+
+  var
+    containsStmt: ContainsStmt
+    deleteStmt: DeleteStmt
+    getStmt: GetStmt
+    putStmt: PutStmt
+
+  if not readOnly:
+    checkExec(env.val, createStmtStr)
+
+    deleteStmt = ? DeleteStmt.prepare(env.val, deleteStmtStr)
+    putStmt = ? PutStmt.prepare(env.val, putStmtStr)
+
+  containsStmt = ? ContainsStmt.prepare(env.val, containsStmtStr)
+  getStmt = ? GetStmt.prepare(env.val, getStmtStr)
+
+  # if a readOnly/existing database does not satisfy the expected schema
+  # `pepare()` will fail and `new` will return an error with message
+  # "SQL logic error"
+
+  success T(dbPath: dbPath, containsStmt: containsStmt, deleteStmt: deleteStmt,
+            env: env.release, getStmt: getStmt, putStmt: putStmt,
+            readOnly: readOnly)
+
+proc dbPath*(self: SQLiteDatastore): string =
+  self.dbPath
+
+proc env*(self: SQLiteDatastore): SQLite =
+  self.env
+
+proc close*(self: SQLiteDatastore) =
+  self.containsStmt.dispose
+  self.getStmt.dispose
+
+  if not self.readOnly:
+    self.deleteStmt.dispose
+    self.putStmt.dispose
+
+  self.env.dispose
+  self[] = SQLiteDatastore()[]
+
+proc timestamp*(t = epochTime()): int64 =
+  (t * 1_000_000).int64
+
+proc idCol*(
+  s: RawStmtPtr,
+  index = 0): string =
+
+  $sqlite3_column_text(s, index.cint).cstring
+
+proc dataCol*(
+  s: RawStmtPtr,
+  index = 1): seq[byte] =
+
+  let
+    i = index.cint
+    dataBytes = cast[ptr UncheckedArray[byte]](sqlite3_column_blob(s, i))
+    dataLen = sqlite3_column_bytes(s, i)
+
+  @(toOpenArray(dataBytes, 0, dataLen - 1))
+
+proc timestampCol*(
+  s: RawStmtPtr,
+  index = 2): int64 =
+
+  sqlite3_column_int64(s, index.cint)
+
+method contains*(
+  self: SQLiteDatastore,
+  key: Key): ?!bool {.locks: "unknown".} =
+
+  var
+    exists = false
+
+  proc onData(s: RawStmtPtr) {.closure.} =
+    let
+      v = sqlite3_column_int64(s, 0)
+
+    if v == 1: exists = true
```

Comment:
should be able to simple covert to boolean with `bool(sqlite3_column_int64(s, 0))`, no need for intermediate `v` here.

Reply:
Good catch, will give it a shot.

Reply:
Nitpicking: sqlite here returns its own representation of bool, which happens to be the same as Nim one. I think it's ok to use just bool(...) since neither one is likely to change, but I find it important to document that - so that reader understands that sqlite bool is returned by EXISTS and converted to Nim bool here

Reply:
I'll add a comment noting the conversion.

Reply:
Done.